src/scheduler/pbs_sched.c

```diff
@@ -791,46 +502,285 @@ are_we_primary()
+		goto rerr;
+	if (dis_flush(fd) != 0)
+		goto rerr;
+	pbs_errno = 0;
+	reply = PBSD_rdrpy(fd);
+	if (reply == NULL)
+		goto rerr;
+	if (pbs_errno != 0) {
+		char *errmsg = get_conn_errtxt(fd);
+		if (errmsg) {
+			log_eventf(PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_NOTICE,
+				   msg_daemonname,
+				   "Server rejected register request for %s connection with error: %s", conn_type, errmsg);
+		}
+		goto rerr;
+	}
+
+	PBSD_FreeReply(reply);
+	return fd;
+
+rerr:
+	pbs_disconnect(fd);
+	PBSD_FreeReply(reply);
+	return -1;
+}
+
+static void
+connect_server(char *svrhost)
+{
+	int pfd = -1;
+	int sfd = -1;
+	svr_t *svr;
+
+	while (pfd < 0 || sfd < 0) {
+		if (pfd < 0) {
+			pfd = connect_server_helper(svrhost, 1);
+			if (pfd < 0) {
+				sleep(2);
+				continue;
+			}
+		}
+		sfd = connect_server_helper(svrhost, 0);
+		if (sfd < 0) {
+			sleep(2);
+			continue;
+		}
+	}
+	tpp_em_add_fd(poll_context, sfd, EM_IN | EM_HUP | EM_ERR);
+
+	svr = malloc(sizeof(svr_t));
+	if (!svr) {
+		log_errf(-1, __func__, "Out of memory while allocating svr struct");
+		exit(1);
+	}
+
+	svr->svrhost = svrhost;
+	svr->pfd = pfd;
+	svr->sfd = sfd;
+	CLEAR_LINK(svr->all_svrs_link);
+	append_link(&servers, &svr->all_svrs_link, svr);
+	set_validate_sched_attrs(svr->pfd);
+	log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO,
+		   msg_daemonname, "Connected to server %s", svrhost ? svrhost : pbs_default());
+}
+
+static void
+connect_servers(void)
+{
+	int max_connection = 1;
+
+	if (poll_context == NULL) {
+		poll_context = tpp_em_init(max_connection);
+		if (poll_context == NULL) {
+			log_errf(-1, __func__, "Failed to init cmd connections context");
+			exit(1);
+		}
+
+		CLEAR_HEAD(servers);
+	}
+
+	connect_server(NULL);
+}
+
+static void
+reconnect_server(svr_t *sconn)
+{
+	char *svrhost = sconn->svrhost;
+
+	close_server(sconn);
+	connect_server(svrhost);
+}
+
+static int
+read_cmd(int sfd, svr_t **sconn)
+{
+	svr_t *cur = NULL, *next;
+	int rc = -1;
+
+	cur = GET_NEXT(servers);
+	while (cur) {
+		next = GET_NEXT(cur->all_svrs_link);
+		if (cur->sfd == sfd)
+			break;
+		cur = next;
+	}
+	if (!cur) /* this shouldn't happen, unless we are completly messed up */
+		return rc;
+
+	if ((rc = get_sched_cmd(cur->sfd, &cur->cmd)) == 1) {
+		if (*sconn == NULL)
+			*sconn = cur;
+	} else {
+		/* force set sconn so wait_for_cmd can use it to reconnect */
+		*sconn = cur;
+	}
+	return rc;
+}
+
+static void
+wait_for_cmd(svr_t **sconn)
+{
+	int nfds;
+	int i;
+	em_event_t *events;
+	int err;
+	int timeout = 10;
+	int hascmd = 0;
+	sigset_t emptyset;
+
+
+again:
+	*sconn = NULL;
+
+	while (!hascmd) {
+		sigemptyset(&emptyset);
+		nfds = tpp_em_pwait(poll_context, &events, timeout, &emptyset);
+		err = errno;
+
+		if (nfds < 0) {
+			if (!(err == EINTR || err == EAGAIN || err == 0)) {
+				log_errf(err, __func__, " tpp_em_wait() error, errno=%d", err);
+				sleep(1); /* wait for 1s for not to burn too much CPU */
```

Comment:
In case if server is down even 1 second sleep also makes the disk full. One thought is we can make this configurable.

Reply:
this is not related to server is down, if server is down then sched will be in connect_server_helper and it will be in pbs_connect() + sleep(2) loop, where we don't have any message.